File: 42. Trapping Rain Water/main.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height)-1
        lmax,rmax = 0,0
        res = 0

        while left < right:
            lmax = max(lmax, height[left])
            rmax = max(rmax, height[right])

            if (lmax < rmax):
                res += (lmax-height[left])
                left+=1
            else:
                res += (rmax-height[right])
                right-=1

        return res

s = Solution()
s.trap([0,1,0,2,1,0,1,3,2,1,2,1])

File: 42. Trapping Rain Water/main.go
package main

func trap(height []int) int {
	l, r := 0, len(height)-1
	lmax, rmax := 0, 0

	res := 0
	for l < r {
		lmax = max(lmax, height[l])
		rmax = max(rmax, height[r])

		if lmax < rmax {
			res += lmax - height[l]
			l++
		} else {
			res += rmax - height[r]
			r--
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
